reps/2022-04-07-serialization.md

```diff
@@ -0,0 +1,300 @@
+
+Let's take `List[bytearray]` as an example to see how to handle nested types. We've already implemented a serializer for `bytearray` above.
+
+Note that in real world we might simply wrap Msgpack to do serialization for common classes. Here we rebuild the wheel only to show how it works.
+
+```python
+class ListSerializer(RaySerializer):
+
+    def serialize(self, list_object: List[bytearray]) -> RaySerializationResult:
+        result = RaySerializationResult()
+        result.in_band_buffer = bytearray()
+        # Total element count.
+        result.in_band_buffer.extend(to_bytes(len(list_object))))
+        for element in list_object:
+            element_res = ray.serialize(element)
+            # in band buffer length
+            result.in_band_buffer.extend(to_bytes(len(element_res.in_band_buffer)))
+            # in band buffer
+            result.in_band_buffer.extend(element_res.in_band_buffer)
+            # append out of band buffer
+            result.out_of_band_buffers += element_res.out_of_band_buffers
+        return result
+
+    def deserialize(self, in_band_buffer: bytes,
+                    oob_iterator: Iterator[memoryview]) -> List[bytearray]:
+        result = []
+        pos = 0
+        # read element count
+        count, pos = get_int(in_band_buffer, pos)
+        for i in range(count):
+            # read element size
+            element_size, pos = get_int(in_band_buffer, pos)
+            # read element in-band data
+            element_in_band_data = in_band_buffer[pos: pos + element_size]
+            pos = pos + element_size
+            # nested serialization here
+            element = ray.deserialize(element_in_band_data, oob_iterator)
+            result.append(element)
+        return result
+```
+
+### Fallback Serializer
+
+By default, if no serializer is registered, use serializer in current Ray's version as the fallback. For example, pickle5 in Python, FST in Java.
+
+We'll register a fallback serializer to Ray. For example, pickle5 fallback serializer will be something like the following, all OOB buffers of pickle5 will be wrapped to Ray's OOB buffer.
+
+```python
+# Pseudo code
+class Pickle5FallbackSerializer(RaySerializer):
+    def serialize(self, instance) -> RaySerializationResult:
+        pickle_oob_buffers = []
+        in_band = pickle5.dumps(instance, buffer_callback=pickle_oob_buffers.append)
+        ray_oob_buffers = convert_to_ray_oob_buffers(pickle_oob_buffers)
+        return RaySerializationResult(in_band, ray_oob_buffers)
+
+    def deserialize(self, in_band_buffer: bytes,
+                    oob_iterator: Iterator[memoryview]):
+        return pickle5.loads(in_band_buffer, buffers=list(oob_iterator))
+```
+
+Nested serialization will still work in this case. When registering a serializer to Ray, we'll also register it to the fallback serializer(with some wrapping).
+
+```python
+# Pseudo code
+def register_serializer(class_name: str, class_type: type, serializer):
```

Comment:
Instead of special casing pickle5, could we add this to the serialization API? Basically, notify each serializer if another serializer is registered after it. That way the pickle5 custom serializer can be notified that ArrowTable, etc. have custom serializers.

Reply:
Or do you want to make it so that only the fallback serializer can see all serializers?

Reply:
> Instead of special casing pickle5, could we add this to the serialization plugin API? 

@ericl Yes I agree with you.

> Or do you want to make it so that only the fallback serializer can see all serializers?

Yes I think only the fallback serializer. Notify all serializers seems like an overkill.
Committed, would you please have a look?

Reply:
It seems good to me. In the ideal case, we would support nesting any kind of serializers, but I can't think of a concrete use case right now, and this also seems fairly easy to add later should we need it.